### DISCUSSION TITLE
WebGL Mujoco Unity

### DISCUSSION THREAD

**Discussion by ALitleLight** [google-deepmind/mujoco#1969] [Asking for Help]
Hi, Mr Balint-H

I'm a student and I'm trying to put mujoco model into unity and build WebGL, I'm now able to synchronize the behavior of Mujoco with the Unity model using mjremote. But when I build WebGL, there are some abnormal behavior. 

Here is a screenshot , illustrating my question:

NullReferenceException: Failed to create Mujoco runtime.
DllNotFoundException: Unable to load DLL 'mjplugin155'. Tried the load the following dynamic libraries: 
![Snipaste_2024-07-23_18-57-55](https://github.com/user-attachments/assets/53691832-f69f-4752-ace7-8b94a4cdb627)

![Snipaste_2024-07-23_18-58-06](https://github.com/user-attachments/assets/7bf78a52-0396-4a05-92df-f52bc251f825)
it may be a simple question,but I've been stuck for a long time and hope you help me solve it. I would be grateful for your answer.

**Comment by Balint-H** [comment]
Hello! WebGL builds usually have a limitation of not being able to use external libraries (such as dll files). However, since MuJoCo's source code is written in C/++, I believe it would be possible to include it in your project and make a build for browsers that way.  Of course if you do include the source code in your project, and intend to distribute it please be aware of MuJoCo's license T&Cs.

For reference:
https://discussions.unity.com/t/include-c-dll-in-webgl-build/607300

However, I suspect some changes to the plugin would be needed to support in-project compilation of MuJoCo, (e.g. [changing the DllImport calls in the bindings](https://github.com/google-deepmind/mujoco/blob/2d24c588195439ae806c4dee97252eab6448a94d/unity/Runtime/Bindings/MjBindings.cs#L6399) according to the [guidance in the Unity docs for referencing internal C/++ libraries](https://docs.unity3d.com/Manual/web-interacting-browsers-c-to-unity.html))

Having support for this would be a nice feature! It is not available now, but I intend to investigate this later on (not this month).

If you are interested in trying to implement this on your own, the first step would be to confirm you can compile a hello-world from C in Unity, then move the relevant MuJoCo source code into your project, and finally to update the bindings (using compiler directives to select the right call based on whether you are running things from editor, making a desktop build or a WebGL one).
